### PR TITLE
[webpack-config-utils]: add type definitions

### PR DESCRIPTION
**Pull request by Hotell**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/webpack-config-utils/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    type EnvVars = 'production' | 'prod' | 'test' | 'development' | 'dev';
+    interface RemoveEmpty {
+        <T>(input: Array<T | undefined>): T[];
+        <T>(input: { [P in keyof T]: T[P] }): NonEmptyObject<T>;
+    }
+    type GetIfUtils = <E extends EnvVars | string>(
+        env: { [P in E]: boolean | string } | E,
+        vars?: Array<EnvVars | string>
+    ) => IfUtils;
+    // @TODO
```
Comment (Hotell) [inline types/webpack-config-utils/index.d.ts]: Hey folks! Any ideas how to use this?

Generics are not properly inferred when I wanna use this conditional mapping

```ts
type Falsy = false | '' | undefined | null | 0;
type PropIf = <A, I, E>(add: A, value: I, alternate: E) => A extends Falsy ? E : I;

//...

 // $ExpectType "string"
propIf(true, 'value', 'alternate'); // 'value'

// if generics are explicitly provided 👉

 // $ExpectType "value"
propIf<true, 'value', 'alternate'>(true, 'value', 'alternate');
```

File: types/webpack-config-utils/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    interface RemoveEmpty {
+        <T>(input: Array<T | undefined>): T[];
+        <T>(input: { [P in keyof T]: T[P] }): NonEmptyObject<T>;
+    }
+    type GetIfUtils = <E extends EnvVars | string>(
+        env: { [P in E]: boolean | string } | E,
+        vars?: Array<EnvVars | string>
+    ) => IfUtils;
+
+    // @TODO
```
Comment (Hotell) [inline types/webpack-config-utils/index.d.ts]: Hey folks! Any ideas how to use this?

Generics are not properly inferred when I wanna use this conditional mapping

```ts
type Falsy = false | '' | undefined | null | 0;
type PropIf = <A, I, E>(add: A, value: I, alternate: E) => A extends Falsy ? E : I;

//...

 // $ExpectType "string"
propIf(true, 'value', 'alternate'); // 'value'

// if generics are explicitly provided 👉

 // $ExpectType "value"
propIf<true, 'value', 'alternate'>(true, 'value', 'alternate');
```

Comment (sandersn) [reply types/webpack-config-utils/index.d.ts]: I'll play around to see if I can get conditional types working the way you want, but in the meantime, here is the old-style overload set that does what you want:

```ts
    interface API {
        getIfUtils: GetIfUtils;
        removeEmpty: RemoveEmpty;
        propIf<E>(a: Falsy, value: any, alternate: E): E
        propIf<I>(a: any, value: I, alternate: any): I
        propIfNot<I>(a: Falsy, value: I, alternate: any): I
        propIfNot<E>(a: any, value: any, alternate: E): E
    }
```

File: types/webpack-config-utils/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        ifProduction: IfUtilsFn;
+        ifNotProduction: IfUtilsFn;
+        ifProd: IfUtilsFn;
+        ifNotProd: IfUtilsFn;
+        ifTest: IfUtilsFn;
+        ifNotTest: IfUtilsFn;
+        [key: string]: IfUtilsFn;
+    }
+
+    interface API {
```
Comment (sandersn) [inline types/webpack-config-utils/index.d.ts]: I'd get rid of the namespace and just `export var getIfUtils: GetIfUtils; export var removeEmpty: RemoveEmpty` etc from the toplevel.

**Review by sandersn — CHANGES_REQUESTED** [review]

**Comment by sandersn** [comment]
The basic problem is that A widens. Even then, the only values that fail are true, false, "" and 0. Here's a miniature example:

```ts
declare function pt<A, T, F>(add: A): A extends Falsy ? 'f' : 't';
```

For `pt(true)`, A widens to `boolean`, which is equivalent to `true | false`. Then `true` does not extend Falsy, resulting in 't', and `false` does, resulting in 'f'. Union the two together and the return type is `'t' | 'f'`. The result is similar for `pt(false)`.

For `pt("hi")`, A widens to string. string doesn't extend Falsy, so you just get 't'. That's correct for "hi", but "" *also* widens to string and *also* incorrectly gives 't'.

Unfortunately, I don't know why A is widening here.

**Comment by Hotell** [comment]
Many thanks for help @sandersn ! One think though. I've removed namespace, with that linter complains if not all ambient types aren't exporter. Is this ok behaviour ? I mean, with this declaration, even private API's are accessible via import for consumer which is confusing. WDYT ?

**Comment by sandersn** [comment]
I can only guess about the reason for the lint rule, maybe it is that DefinitelyTyped is for documenting the public interface of a package, so a function that uses a private type will only frustrate people?

I'm not sure whether this makes sense for this particular case though. Which types did you not intend to export?

**Review by sandersn — APPROVED** [review]

**Comment by Hotell** [comment]
Well only public API

```ts
export function getIfUtils<E extends EnvVars | string>(
    env: { [P in E]: boolean | string } | E,
    vars?: Array<EnvVars | string>
): IfUtils;
export function removeEmpty<T>(input: Array<T | undefined>): T[];
export function removeEmpty<T>(input: { [P in keyof T]: T[P] }): NonEmptyObject<T>;
export function propIf<E>(a: Falsy, value: any, alternate: E): E;
export function propIf<I>(a: any, value: I, alternate: any): I;
export function propIfNot<I>(a: Falsy, value: I, alternate: any): I;
export function propIfNot<E>(a: any, value: any, alternate: E): E;
```

the rest is private. But the linter throws errors

so I can either:
- turn off that lint rule
- wrap it back to `declare namespace`  and export via old syntax `export = myNamespace.Api`

I know that prior to TS 2.9 this was needed when user wanted to generate type definitions from source, as compiler was unable to get those private types, which is fixed in 2.9 thanks to `import(...)`

The lint rule might be related to that maybe?

**Comment by sandersn** [comment]
Oh. I *think* I see, but I think I might agree with the linter. For example, since IfUtils is exported by getIfUtils, shouldn’t it also be accessible? Otherwise we will have to show the user a structural type in intellisense. But that’s probably not what you want.

**Comment by paulvanbrenk** [comment]
@sandersn can you merge this if it's ready

**Comment by Hotell** [comment]
@sandersn 

> Oh. I think I see, but I think I might agree with the linter. For example, since IfUtils is exported by getIfUtils, shouldn’t it also be accessible? Otherwise we will have to show the user a structural type in intellisense. But that’s probably not what you want.

I know and I agree as well, I mean TS 2.9 knows these types even if they are not exported, which allows us to provide "real" public API types.

But no worries we can merge this for now ✌️

**Pull request merged**